src/data-access/repositories/user-repository.ts

```diff
@@ -0,0 +1,197 @@
+import { UserInput } from '../../graphql/types/schema-types';
+import { UserModel, UserPageModel } from '../models/user';
+import AWS from 'aws-sdk';
+import { UserInputError } from 'apollo-server-errors';
+import moment from 'moment';
+import validator from 'validator';
+import { v4 as uuid } from 'uuid';
+import { ApolloError } from 'apollo-server-lambda';
+import { DocumentClient, QueryOutput, ScanOutput } from 'aws-sdk/clients/dynamodb';
+import { Maybe } from 'graphql/jsutils/Maybe';
+import { UsersDBConfiguration } from '../../configuration/users-db';
+
+export interface UserRepository {
+	createUser(data: UserInput): Promise<UserModel>;
+	updateUser(id: string, data: UserInput): Promise<UserModel>;
+	deleteUser(id: string): Promise<UserModel>;
+	getUser(id: string): Promise<UserModel>;
+	listUsers(query: Maybe<string>, limit: number, cursor: Maybe<string>): Promise<UserPageModel>;
+}
+
+interface ValidatedInput extends UserInput {
+	formattedDateOfBirth: string | null;
+}
+
+export class DynamoDBUserRepository implements UserRepository {
+	constructor(private database: AWS.DynamoDB.DocumentClient, private config: UsersDBConfiguration) {}
+
+	async createUser(data: UserInput): Promise<UserModel> {
+		if (!data.name) {
+			throw new UserInputError('The name must be specified');
+		}
+
+		if (!data.dob) {
+			throw new UserInputError('The date of birth must be specified');
+		}
+
+		if (!data.address) {
+			throw new UserInputError('The address must be specified');
+		}
+
+		const validatedInput = this.validateInput(data);
+		const user = {
+			id: uuid(),
+			name: data.name,
+			address: data.address,
+			description: data.description,
+			dob: validatedInput.formattedDateOfBirth,
+			createdAt: moment().toISOString(),
+		} as UserModel;
+
+		await this.database
+			.put({
+				TableName: this.config.tableName,
+				Item: user,
+			})
+			.promise();
+
+		return user;
+	}
+
+	async updateUser(id: string, data: UserInput): Promise<UserModel> {
+		const validatedInput = this.validateInput(data);
+		const user = await this.getUser(id);
+
+		user.name = data.name || user.name;
+		user.address = data.address || user.address;
+		user.dob = validatedInput.formattedDateOfBirth || user.dob;
+		user.description = data.description || user.description;
+		user.updatedAt = moment().toISOString();
```

Comment:
What if `data.description` is `''`? 
We wouldn't be able to clear the description, what would be the solution for this?

Reply:
Oh, I didn't think of that. I was about to say I could use a ternary operator for this, but your question made me look up if typescript had any type of null coalescing operator, and turned out it has! So I would go with something like
`data.description ?? user.description`